### PR TITLE
[CI, enhancement] Add ability to build with ```gcov```,  adding C++ code coverage for the ```onedal``` folder in codecov

### PR DESCRIPTION
**Pull request by icfaust**
## Description

Coverage statistics have been recently added to sklearnex via codecov.  This was focused on the Python code, but C++ is the other important aspect for integrating oneDAL (representing a fifth of the code in the repo). This PR does the following:

1) Adds ability to build with gcov for GNU C++ and Intel DPC++ compilers in cmake set by the SKLEARNEX_GCOV environment variable
2) Integrates C++ code coverage statistics in the Windows and Linux DPC++ Github Actions (currently Python 3.9 and Python3.11)
3) Switches coverage.py data format from json format to lcov (with branching enabled) to match C++ addition

It does this by using the ```gcovr``` package, and by a bash script (```generate_coverage_files.sh```). Note, to use gcov, the build Numpy must be installed, otherwise it will fail (usually changed in the requirements-test.txt installation).  Therefore, the bash script uses a ```NUMPY_BUILD``` environment variable in order to store this information. This does not impact the daal4py build, as it is considered legacy code.

Consequences:
1) Adds C++ coverage metrics
2) Greatly reduces coverage percentages (by using branch rather than line coverage which is stricter)
3) Adds ~3 minute total via the longer builds and report generation, increasing the CI runtime by 10%
4) lcov files are now to be considered the standard code coverage format for sklearnex

Note: windows visual studio does not easily maintain a code coverage tool, and is therefore neglected.  The majority of code is operated with the DPC build, with only a small segment being missed.

No performance benchmarks necessary.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: scripts/CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +125,18 @@ if(IFACE STREQUAL "host")
         endif()
     endif()
 
+    if(SKLEARNEX_GCOV)
+        if(CMAKE_CXX_COMPILER_ID STREQUAL "IntelLLVM" AND WIN32)
+            set(CMAKE_CXX_FLAGS "/clang:--coverage ${CMAKE_CXX_FLAGS}")
+            list(APPEND ONEDAL_LIBRARIES "clang_rt.profile-x86_64.lib")
+        elseif(CMAKE_CXX_COMPILER_ID STREQUAL "IntelLLVM" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
```
Comment (david-cortes-intel) [inline scripts/CMakeLists.txt:132]: * What about clang?
* What happens if you invoke the compiler through `ccache`?

Comment (david-cortes-intel) [reply scripts/CMakeLists.txt:132]: If we know that it will work out with clang, I think there's no harm in adding it here. Someone might try to generate these reports locally too.

Comment (icfaust) [reply scripts/CMakeLists.txt]: Done! Generalized to work with the various compile id's which are clang-based.

**Comment by david-cortes-intel** [comment]
@icfaust What would be the right way to generate a report locally? I haven't been able to get an interactive html report from this PR.

**Comment by icfaust** [comment]
> @icfaust What would be the right way to generate a report locally? I haven't been able to get an interactive html report from this PR.

You should be able to use ```generate_coverage_reports.sh``` or most of the commands in there. The big thing is the gcov is touchy with cmake, and the build directory is the root for gcovr.  If you are separating the build install and the test requirements installs, make sure to be careful with the numpy version.

**Comment by david-cortes-intel** [comment]
> > @icfaust What would be the right way to generate a report locally? I haven't been able to get an interactive html report from this PR.
> 
> You should be able to use `generate_coverage_reports.sh` or most of the commands in there. The big thing is the gcov is touchy with cmake, and the build directory is the root for gcovr. If you are separating the build install and the test requirements installs, make sure to be careful with the numpy version. Are you using g++ or icpx?

Meaning: it's not generated as part of the `run.sh` files for coverage? If that's the case, would be ideal to start documenting these things in the .md files.

**Comment by david-cortes-intel** [comment]
@icfaust Can you attach here an .html report of the C++ code coverage that would be generated?

**Comment by icfaust** [comment]
> @icfaust Can you attach here an .html report of the C++ code coverage that would be generated?

I recommend looking at codecov: https://github.com/uxlfoundation/scikit-learn-intelex/pull/2249#issuecomment-2579588945

File: .github/scripts/generate_coverage_reports.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    FILTER=$(realpath ./onedal).*
+    echo $FILTER
+    
+    NUMPY_TEST=$(python -m pip freeze | grep numpy)
+    # install dependencies
+    # proper operation of gcov with sklearnex requires the header files from
+    # the build numpy, this must be previously set as NUMPY_BUILD
+    python -m pip install gcovr $NUMPY_BUILD
+    
+    gcovr --gcov-executable "${GCOV_EXE}" -r . -v --lcov --filter "${FITLER}" -o coverage_cpp_"${1}".info
```
Comment (david-cortes-intel) [inline .github/scripts/generate_coverage_reports.sh:42]:
```suggestion
    gcovr --gcov-executable "${GCOV_EXE}" -r . -v --lcov --filter "${FILTER}" -o coverage_cpp_"${1}".info
```

Comment (icfaust) [reply .github/scripts/generate_coverage_reports.sh:42]: Oh wow, not sure how I didn't pick that up.

**Review by david-cortes-intel — APPROVED** [review]
